### PR TITLE
[Cortex-A]add gic&gtimer interface

### PR DESCRIPTION
**Pull request by fenghuijie**
## 拉取/合并请求描述：(PR description)

[
add gic&gtimer interface
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [x] 本拉取/合并符合[RT-Thread代码规范](../documentation/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/coding_style_en.txt)

### CODE REVIEW

File: libcpu/arm/cortex-a/gic.c [hunk 1]
```diff
@@ -17,6 +17,10 @@
 #include "gic.h"
 #include "cp15.h"
 
+#ifdef RT_USING_FINSH
```
Comment (BernardXiong) [inline libcpu/arm/cortex-a/gic.c:20]: It is not necessary to include this header file.

File: libcpu/arm/cortex-a/gic.c [hunk 2]
```diff
@@ -126,70 +236,140 @@ void arm_gic_set_cpu(rt_uint32_t index, int irq, unsigned int cpumask)
 
     old_tgt = GIC_DIST_TARGET(_gic_table[index].dist_hw_base, irq);
 
-    old_tgt &= ~(0x0FFUL << ((irq % 4)*8));
-    old_tgt |=   cpumask << ((irq % 4)*8);
+    old_tgt &= ~(0x0FF << ((irq % 4)*8));
```
Comment (BernardXiong) [inline libcpu/arm/cortex-a/gic.c:239]: Are you sure, there is no warning?

File: libcpu/arm/cortex-a/gic.c [hunk 3]
```diff
@@ -265,36 +445,50 @@ int arm_gic_cpu_init(rt_uint32_t index, rt_uint32_t cpu_base)
...
+    rt_kprintf("\n");
 }
 
-#ifdef RT_USING_SMP
-void rt_hw_ipi_send(int ipi_vector, unsigned int cpu_mask)
- {
-     /* note: ipi_vector maybe different with irq_vector */
-     GIC_DIST_SOFTINT(_gic_table[0].dist_hw_base) = (cpu_mask << 16) | ipi_vector;
-}
+#ifdef RT_USING_FINSH
```
Comment (BernardXiong) [inline libcpu/arm/cortex-a/gic.c:491]: It's better to use msh command for debug and MSH_CMD_EXPORT will be defined as empty if not using finsh/msh. Therefore, we can remove the `#ifdef RT_USING_FINSH` condition.

File: libcpu/arm/cortex-a/interrupt.c [hunk 4]
```diff
@@ -83,7 +83,7 @@ void rt_hw_interrupt_umask(int vector)
  */
 int rt_hw_interrupt_get_irq(void)
 {
-    return arm_gic_get_active_irq(0) & GIC_ACK_INTID_MASK;
+    return arm_gic_get_active_irq(0);
```
Comment (BernardXiong) [inline libcpu/arm/cortex-a/interrupt.c:86]: Are you sure to remove `GIC_ACK_INTID_MASK`?

**Review by BernardXiong — CHANGES_REQUESTED** [review]
Please follow the comments.

File: libcpu/arm/cortex-a/interrupt.c (hunk 4 above)
Comment (fenghuijie) [reply libcpu/arm/cortex-a/interrupt.c:86]: 这个需要改，不然在多核下会影响核间中断响应

**Pull request merged**